common.c

```diff
@@ -105,3 +114,76 @@ const Instr_t Factorial[PROGRAM_SIZE] = {
     Instr_Print,    // n
     Instr_Halt
 };
+
+cpu_t init_cpu () {
+    cpu_t cpu = {.pc = 0, .sp = -1, .state = Cpu_Running,
+                 .steps = 0, .stack = {0},
+                 .pmem = LoadedProgram ? LoadedProgram : DefProgram};
+    return cpu;
+}
+
+static const char *steplimit_opt = "--steplimit=";
+static const char *inp_prog_opt = "--inp-prog=";
+
+static inline
+void help_dump(char * exec_name, int ret_code) {
+    fprintf(stderr, "Usage: %s %s<num> %s<str>\n", exec_name, steplimit_opt, inp_prog_opt);
+    exit (2);
+}
+
+long long parse_args(int argc, char** argv) {
+    long long steplimit = LLONG_MAX;
+    FILE *prog_file = NULL;
+
+    for (int i = 1; i < argc; ++i) {
+        if (!strcmp(argv[i], "--help"))
+            help_dump(argv[0], 2);
+        else if (!strncmp(argv[i], steplimit_opt, strlen(steplimit_opt))) {
+            char *endptr = NULL;
+            steplimit = strtoll(argv[i] + strlen(steplimit_opt), &endptr, 10);
+            if (errno || (*endptr != '\0')) {
+                fprintf(stderr, "Unrecognized steplimit: %s\a", argv[i]);
+                help_dump(argv[0], 2);
+            }
+        }
+        else if (!strncmp(argv[i], inp_prog_opt, strlen(inp_prog_opt))) {
+            prog_file = fopen(argv[i] + strlen(inp_prog_opt), "rb");
+            if (errno || prog_file == NULL) {
+                fprintf(stderr, "Unrecognized program file: %s\n", argv[i]);
+                help_dump(argv[0], 2);
+            }
+        }
+        else {
+            fprintf(stderr, "Unrecognized option: %s\n", argv[i]);
+            help_dump(argv[0], 2);
+        }
+    }
+
+    unsigned long long filelen = 0;
+    if (prog_file != NULL) {
+        fseek(prog_file, 0, SEEK_END);  // Jump to the end of the file
+        filelen = ftell(prog_file);     // Get the current byte offset in the file
+        rewind(prog_file);              // Jump back to the beginning of the file
+        if (filelen > PROGRAM_SIZE * sizeof(Instr_t))
+            fprintf(stderr, "Warning: input program size exceeds allocated memory.\n");
```

Comment:
In this case, the program has to either fail and exit, or at least truncate the input file to fit into PROGRAM_SIZSE.

The general rule is: if it becomes  known that input parameters will lead to incorrect (undefined for C) state/behavior, a corrective/preventing action *must* be taken/